query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+        schema.schema_map.insert(e, a);
+    }
+
+    #[test]
+    fn test_unknown_ident() {
+        let mut cc = ConjoiningClauses::new();
+        let schema = Schema::new();
+
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?x"))),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Constant(NonIntegerConstant::Boolean(true)),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        assert!(cc.is_known_empty);
+    }
+
+    #[test]
+    fn test_unknown_attribute() {
+        let mut cc = ConjoiningClauses::new();
+        let mut schema = Schema::new();
+
+        associate_ident(&mut schema, ":foo/bar", 99);
+
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(Variable(PlainSymbol::new("?x"))),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Constant(NonIntegerConstant::Boolean(true)),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        assert!(cc.is_known_empty);
+    }
+
+    #[test]
+    fn test_apply_simple_pattern() {
+        let mut cc = ConjoiningClauses::new();
+        let mut schema = Schema::new();
+
+        associate_ident(&mut schema, ":foo/bar", 99);
+        add_attribute(&mut schema, 99, Attribute {
+            value_type: ValueType::Boolean,
+            ..Default::default()
+        });
+
+        let x = Variable(PlainSymbol::new("?x"));
+        cc.apply_pattern(&schema, &Pattern {
+            source: None,
+            entity: PatternNonValuePlace::Variable(x.clone()),
+            attribute: PatternNonValuePlace::Ident(NamespacedKeyword::new("foo", "bar")),
+            value: PatternValuePlace::Constant(NonIntegerConstant::Boolean(true)),
+            tx: PatternNonValuePlace::Placeholder,
+        });
+
+        // println!("{:#?}", cc);
+
+        let d0_e = QualifiedAlias("datoms0".to_string(), DatomsColumn::Entity);
+        let d0_a = QualifiedAlias("datoms0".to_string(), DatomsColumn::Attribute);
+        let d0_v = QualifiedAlias("datoms0".to_string(), DatomsColumn::Value);
+
+        // After this, we know a lot of things:
+        assert!(!cc.is_known_empty);
+        assert_eq!(cc.from, vec![SourceAlias(DatomsTable::Datoms, "datoms0".to_string())]);
+
+        // ?x must be a ref.
+        assert_eq!(cc.known_types.get(&x).unwrap(), &ValueType::Ref);
+
+        // ?x is bound to datoms0.e.
+        assert_eq!(cc.bindings.get(&x).unwrap(), &vec![d0_e.clone()]);
+
+        // Our 'where' clauses are two:
+        // - datoms0.a = 99
+        // - datoms0.v = true
+        // No need for a type tag constraint, because the attribute is known.
+        assert_eq!(cc.wheres, vec![
```

Comment:
They're ordered because predictability is our main weapon for query optimization. The clauses are quite consistent.